pallets/runtime/tests/src/identity_test.rs

```diff
@@ -521,15 +594,280 @@ fn frozen_secondary_keys_cdd_verification_test_we() {
+
+        let bob_key = SecondaryKey::new(signer, perms);
+        let key_with_auth = SecondaryKeyWithAuth {
+            auth_signature,
+            secondary_key: bob_key.into(),
+        };
+        Identity::add_secondary_keys_with_authorization(
+            alice.origin(),
+            vec![key_with_auth],
+            expires_at,
+        )
+    };
+
+    let perm1 = Permissions::empty();
+    let perm2 = Permissions::from_pallet_permissions(vec![PalletPermissions::entire_pallet(
+        b"identity".into(),
+    )]);
+
+    // count alice's secondary keys.
+    let count_keys = || get_secondary_keys(alice.did).len();
+
+    // Add bob's identity signatory with empty permissions
+    let res = add_secondary_key_with_auth(bob_identity_signer, perm1.clone());
+    assert_ok!(res);
+    assert_eq!(count_keys(), 1);
+
+    // Add bob's identity signatory again with non-empty permissions
+    let res = add_secondary_key_with_auth(bob_identity_signer, perm2.clone());
+    assert_noop!(res, Error::AlreadyLinked);
+    assert_eq!(count_keys(), 1);
+
+    // Add bob's identity signatory again.
+    let res = add_secondary_key_with_auth(bob_identity_signer, perm1.clone());
+    assert_noop!(res, Error::AlreadyLinked);
+    assert_eq!(count_keys(), 1);
+}
+
+#[test]
+fn join_identity_as_identity_with_perm_test() {
+    ExtBuilder::default()
+        .monied(true)
+        .build()
+        .execute_with(&do_join_identity_as_identity_with_perm_test);
+}
+
+fn do_join_identity_as_identity_with_perm_test() {
+    let bob = User::new(AccountKeyring::Bob);
+    let bob_identity_signer = Signatory::Identity(bob.did);
+    let alice = User::new(AccountKeyring::Alice);
+
+    // Use `add_auth` and `join_identity` to add a secondary key.
+    let join_identity_with_perms = |signer, perms| {
+        let auth_id = Identity::add_auth(
+            alice.did,
+            signer,
+            AuthorizationData::JoinIdentity(perms),
+            None,
+        );
+        Identity::join_identity(signer, auth_id)
+    };
+
+    let perm1 = Permissions::empty();
+    let perm2 = Permissions::from_pallet_permissions(vec![PalletPermissions::entire_pallet(
+        b"identity".into(),
+    )]);
+
+    // count alice's secondary keys.
+    let count_keys = || get_secondary_keys(alice.did).len();
+
+    // Add bob's identity signatory with empty permissions
+    let res = join_identity_with_perms(bob_identity_signer, perm1.clone());
+    assert_ok!(res);
+    assert_eq!(count_keys(), 1);
+
+    // Add bob's identity signatory again with non-empty permissions
+    let res = join_identity_with_perms(bob_identity_signer, perm2.clone());
+    assert_noop!(res, Error::AlreadyLinked);
+    assert_eq!(count_keys(), 1);
+
+    // Add bob's identity signatory again.
+    let res = join_identity_with_perms(bob_identity_signer, perm1.clone());
+    assert_noop!(res, Error::AlreadyLinked);
+    assert_eq!(count_keys(), 1);
```

Comment:
This duplicates a lot of logic from above. Let's use a higher order function?

Reply:
Sorry, but I don't have time tonight to rewrite and merge the logic of those two tests.

Reply:
Alright, can you follow up at a later time?